File: Engine/runtime_tests.ts
import { downloadArchive, isCompatibleVersion, resolveBrowser, runSystemCheck } from "./runtime.ts";
import { RequestError } from "./credentials.ts";

function assert(value: unknown, message = "Assertion failed"): asserts value {
  if (!value) throw new Error(message);
}

async function missing(path: string): Promise<boolean> {
  try { await Deno.stat(path); return false; }
  catch (error) { if (error instanceof Deno.errors.NotFound) return true; throw error; }
}

async function fixture(handler: () => Response, test: (url: string, directory: string) => Promise<void>): Promise<void> {
  const directory = await Deno.makeTempDir({ prefix: "aster-runtime-test-" });
  const server = Deno.serve({ hostname: "127.0.0.1", port: 0, onListen: () => {} }, handler);
  try { await test(`http://127.0.0.1:${server.addr.port}`, directory); }
  finally {
    await server.shutdown();
    await Deno.remove(directory, { recursive: true });
  }
}

Deno.test("system check returns stdout without adding command diagnostics", async () => {
  const result = await runSystemCheck("Test check", "/bin/sh", ["-c", "printf '152.0.7977.82\\n'"], new AbortController().signal);
  assert(result === "152.0.7977.82");
});

Deno.test("system check failures identify the step and exit code without exposing command output", async () => {
  const result = await runSystemCheck("Chromium signature check", "/bin/sh", ["-c", "printf 'private stdout'\nprintf 'private stderr' >&2\nexit 7"], new AbortController().signal)
    .then(() => undefined, (error) => error);
  assert(result instanceof RequestError && result.code === "browser_verify");
  assert(result.message === "Chromium signature check failed (sh, exit 7).");
});

Deno.test("system check timeout stops and reaps the process and identifies the step", async () => {
  const result = await runSystemCheck("Chromium Gatekeeper check", "/bin/sleep", ["10"], new AbortController().signal, 25)
    .then(() => undefined, (error) => error);
  assert(result instanceof RequestError && result.code === "browser_setup");
  assert(result.message === "Chromium Gatekeeper check timed out (sleep).");
});

Deno.test("system check cancellation stops and reaps the process", async () => {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 25);
  try {
    const result = await runSystemCheck("Chromium Gatekeeper check", "/bin/sleep", ["10"], controller.signal)
      .then(() => undefined, (error) => error);
    assert(result instanceof RequestError && result.code === "cancelled");
  } finally { clearTimeout(timer); }
});

Deno.test("browser versions use numeric components and reject malformed or older builds", () => {
  for (const version of ["152.0.7977.82", "152.0.7977.83", "152.0.7978.0", "153.0.0.0"]) {
    assert(isCompatibleVersion(version), `Compatible version rejected: ${version}`);
  }
  for (const version of ["151.9.9999.999", "152.0.7977.9", "152.0.7977.81", "152.0.7977", "152.0.7977.82-beta", " 152.0.7977.82", "99999999999999999.0.0.0", "152.0.7977.8e2"]) {
    assert(!isCompatibleVersion(version), `Invalid version accepted: ${version}`);
  }
});

Deno.test("verified downloads retain the exact bytes in a private file", async () => {
  await fixture(() => new Response("abc"), async (url, directory) => {
    const file = `${directory}/browser.dmg`;
    await downloadArchive(url, file, "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad", new AbortController().signal, () => {});
    assert(await Deno.readTextFile(file) === "abc");
    assert(((await Deno.stat(file)).mode! & 0o777) === 0o600, "The archive is not private");
  });
});

Deno.test("a checksum failure removes the downloaded file", async () => {
  await fixture(() => new Response("unexpected archive"), async (url, directory) => {
    const file = `${directory}/browser.dmg`;
    const result = await downloadArchive(url, file, "0".repeat(64), new AbortController().signal, () => {}).then(() => undefined, (error) => error);
    assert(result instanceof RequestError && result.code === "browser_checksum");
    assert(await missing(file), "Checksum failure left an archive");
  });
});

Deno.test("download cancellation removes a partial file and closes its response", async () => {
  let streamCancelled = false;
  await fixture(() => new Response(new ReadableStream({
    start(controller) { controller.enqueue(new TextEncoder().encode("partial archive")); },
    cancel() { streamCancelled = true; }
  })), async (url, directory) => {
    const file = `${directory}/browser.dmg`;
    const controller = new AbortController();
    const result = await downloadArchive(url, file, "0".repeat(64), controller.signal, () => controller.abort()).then(() => undefined, (error) => error);
    assert(result instanceof RequestError && result.code === "cancelled");
    assert(await missing(file), "Cancellation left a partial archive");
  });
  assert(streamCancelled, "Cancellation left the HTTP response open");
});

Deno.test("an HTTP failure creates no archive", async () => {
  await fixture(() => new Response("unavailable", { status: 503 }), async (url, directory) => {
    const file = `${directory}/browser.dmg`;
    const result = await downloadArchive(url, file, "0".repeat(64), new AbortController().signal, () => {}).then(() => undefined, (error) => error);
    assert(result instanceof RequestError && result.code === "browser_download");
    assert(await missing(file));
  });
});

Deno.test("download refuses to replace a file it does not own", async () => {
  await fixture(() => new Response("abc"), async (url, directory) => {
    const file = `${directory}/browser.dmg`;
    await Deno.writeTextFile(file, "existing file");
    const result = await downloadArchive(url, file, "0".repeat(64), new AbortController().signal, () => {}).then(() => undefined, (error) => error);
    assert(result instanceof Deno.errors.AlreadyExists);
    assert(await Deno.readTextFile(file) === "existing file");
  });
});

Deno.test("cancelled setup creates no cache or staging directory", async () => {
  const directory = await Deno.makeTempDir({ prefix: "aster-runtime-test-" });
  try {
    const controller = new AbortController();
    controller.abort();
    const result = await resolveBrowser(directory, controller.signal, () => { throw new Error("Cancelled setup reported progress"); }).then(() => undefined, (error) => error);
    assert(result instanceof RequestError && result.code === "cancelled");
    const entries = [];
    for await (const entry of Deno.readDir(directory)) entries.push(entry.name);
    assert(entries.length === 0, "Cancelled setup left files");
  } finally { await Deno.remove(directory, { recursive: true }); }
});

File: Tests/Engine/RuntimeTests.swift
import Darwin
import Foundation
import Network

@MainActor
private func runtimeExpect(_ condition: Bool, _ message: String) throws {
    if !condition { throw EngineFailure("test", message) }
}

@MainActor
private func runtimeFailure(_ code: String, _ operation: () async throws -> Void) async throws {
    do { try await operation() }
    catch let failure as EngineFailure {
        try runtimeExpect(failure.code == code, "Unexpected runtime error code: \(failure.code)")
        return
    }
    throw EngineFailure("test", "Expected runtime error \(code)")
}

@MainActor
private func runtimeUntil(_ condition: () -> Bool) async throws {
    for _ in 0..<500 {
        if condition() { return }
        try await Task.sleep(for: .milliseconds(10))
    }
    throw EngineFailure("test", "Runtime fixture timed out.")
}

@MainActor
private final class RuntimeHTTPFixture {
    private let listener: NWListener
    private var clients: [NWConnection] = []
    private var startup: CheckedContinuation<URL, any Error>?

    init(status: Int = 200, body: String = "abc", hold: Bool = false) throws {
        let parameters = NWParameters.tcp
        parameters.requiredLocalEndpoint = .hostPort(host: "127.0.0.1", port: .any)
        listener = try NWListener(using: parameters, on: .any)
        listener.stateUpdateHandler = { [weak self] state in
            MainActor.assumeIsolated {
                guard let self else { return }
                if case .ready = state, let port = self.listener.port {
                    self.startup?.resume(returning: URL(string: "http://127.0.0.1:\(port.rawValue)/archive")!)
                    self.startup = nil
                } else if case .failed = state {
                    self.startup?.resume(throwing: EngineFailure("test", "HTTP fixture did not start."))
                    self.startup = nil
                }
            }
        }
        listener.newConnectionHandler = { [weak self] connection in
            MainActor.assumeIsolated {
                self?.clients.append(connection)
                connection.start(queue: .main)
                connection.receive(minimumIncompleteLength: 1, maximumLength: 8_192) { _, _, _, _ in
                    let length = hold ? 1_000_000 : body.utf8.count
                    let response = "HTTP/1.1 \(status) Test\r\nContent-Type: application/octet-stream\r\nContent-Length: \(length)\r\nConnection: close\r\n\r\n\(body)"
                    connection.send(content: Data(response.utf8), contentContext: hold ? .defaultMessage : .finalMessage,
                                    isComplete: true, completion: .contentProcessed { _ in })
                }
            }
        }
    }

    func start() async throws -> URL {
        try await withCheckedThrowingContinuation {
            startup = $0
            listener.start(queue: .main)
        }
    }

    func stop() {
        listener.cancel()
        for client in clients { client.cancel() }
        clients.removeAll()
    }
}

@MainActor
private func runtimeDownloadTest(status: Int = 200, body: String = "abc", hold: Bool = false,
                                 _ test: (URL, URL) async throws -> Void) async throws {
    let directory = try BrowserRuntime.makePrivateDirectory(in: FileManager.default.temporaryDirectory, prefix: "passtrami-runtime-test-")
    defer { try? FileManager.default.removeItem(at: directory) }
    let server = try RuntimeHTTPFixture(status: status, body: body, hold: hold)
    defer { server.stop() }
    try await test(server.start(), directory.appendingPathComponent("browser.dmg"))
}

@MainActor
private final class RuntimeDownloadTask {
    var task: Task<Void, any Error>?
}

@MainActor
private func runtimeBrowserTest(mode: String) async throws {
    let manager = FileManager.default
    let root = try BrowserRuntime.makePrivateDirectory(in: manager.temporaryDirectory, prefix: "passtrami-browser-test-")
    defer { try? manager.removeItem(at: root) }
    let data = root.appendingPathComponent("data")
    let resources = root.appendingPathComponent("resources")
    for directory in [data, resources.appendingPathComponent("AppleExtension"), resources.appendingPathComponent("Engine")] {
        try manager.createDirectory(at: directory, withIntermediateDirectories: true)
    }
    try Data("// Fixture extension\n".utf8).write(to: resources.appendingPathComponent("AppleExtension/background.js"))
    try Data("// Fixture bridge\n".utf8).write(to: resources.appendingPathComponent("Engine/bridge.js"))
    let quote = { (value: String) in "'" + value.replacingOccurrences(of: "'", with: "'\\''") + "'" }
    let pidFile = root.appendingPathComponent("pid")
    let commandFile = root.appendingPathComponent("command")
    let argumentsFile = root.appendingPathComponent("arguments")
    let leakedFile = root.appendingPathComponent("leaked-fd")
    let null = Darwin.open("/dev/null", O_RDWR)
    defer { Darwin.close(null) }
    let unrelated = fcntl(null, F_DUPFD, 80)
    try runtimeExpect(unrelated >= 80, "Could not create an unrelated descriptor.")
    defer { Darwin.close(unrelated) }
    let executable = root.appendingPathComponent("browser")
    let readCommand = mode == "wait-ready" ? "" : "IFS= read -r -d '' command <&3\nprintf '%s' \"$command\" > \(quote(commandFile.path))\n"
    let response: String
    switch mode {
    case "ready":
        // Exercise multiple frames and a reply split across separate writes.
        response = "printf '{\"method\":\"fixture.event\"}\\0{\"id\":1,\"result\":' >&4\n/bin/sleep 0.02\nprintf '{\"id\":\"fixture\"}}\\0' >&4\n"
    case "invalid": response = "printf 'invalid JSON\\0' >&4\n"
    case "oversized": response = "/usr/bin/head -c 1048577 /dev/zero | /usr/bin/tr '\\000' x >&4\n"
    case "error": response = "printf '{\"id\":1,\"error\":{\"code\":-1}}\\0' >&4\n"
    case "exit": response = "exit 0\n"
    default: response = ""
    }
    let script = """
    #!/bin/bash
    printf '%s' "$$" > \(quote(pidFile.path))
    printf '%s\\n' "$@" > \(quote(argumentsFile.path))
    if { : <&\(unrelated); } 2>/dev/null; then : > \(quote(leakedFile.path)); fi
    \(readCommand)\(response)exec /bin/sleep 30
    """
    try Data((mode == "spawn-error" ? "Invalid executable fixture" : script).utf8).write(to: executable)
    try manager.setAttributes([.posixPermissions: 0o700], ofItemAtPath: executable.path)
    var exits = 0
    let startup = Task {
        try await BrowserSession.start(executable: executable, resources: resources, dataDirectory: data,
                                       port: 1, token: "fixture-token", onExit: { exits += 1 })
    }
    var browser: BrowserSession?
    do {
        if mode == "spawn-error" {
            do {
                _ = try await startup.value
                throw EngineFailure("test", "An invalid executable was accepted.")
            } catch is POSIXError { }
            try runtimeExpect(try manager.contentsOfDirectory(atPath: data.path).isEmpty, "Spawn failure left a profile.")
            return
        }
        try await runtimeUntil { (try? String(contentsOf: pidFile, encoding: .utf8)).flatMap(Int32.init) != nil }
        let pid = Int32(try String(contentsOf: pidFile, encoding: .utf8))!
        if mode == "ready" {
            browser = try await startup.value
            let command = try JSONSerialization.jsonObject(with: Data(contentsOf: commandFile)) as! [String: Any]
            try runtimeExpect(command["method"] as? String == "Extensions.loadUnpacked", "Browser did not send the CDP load command.")
            let arguments = try String(contentsOf: argumentsFile, encoding: .utf8).split(separator: "\n")
            try runtimeExpect(arguments.contains("--remote-debugging-pipe") && !arguments.contains(where: { $0.hasPrefix("--remote-debugging-port") }),
                              "Browser did not use an exclusive debugging pipe.")
            try runtimeExpect(!manager.fileExists(atPath: leakedFile.path), "Browser inherited an unrelated descriptor.")
            let sessions = try manager.contentsOfDirectory(at: data, includingPropertiesForKeys: nil)
            try runtimeExpect(sessions.count == 1, "Browser created extra sessions.")
            let preferences = try JSONSerialization.jsonObject(with: Data(contentsOf: sessions[0].appendingPathComponent("profile/Default/Preferences"))) as! [String: Any]
            try runtimeExpect(preferences["password_manager_enabled"] as? Bool == false, "Browser enabled its own password manager.")
            let original = try String(contentsOf: resources.appendingPathComponent("AppleExtension/background.js"), encoding: .utf8)
            try runtimeExpect(original == "// Fixture extension\n", "Browser changed the source extension.")
            Darwin.kill(pid, SIGTERM)
            try await runtimeUntil { exits == 1 }
            await browser?.stop()
        } else {
            if mode == "wait-ready" || mode == "wait-cdp" {
                if mode == "wait-cdp" { try await runtimeUntil { ((try? Data(contentsOf: commandFile))?.count ?? 0) > 0 } }
                startup.cancel()
                try await runtimeFailure("cancelled") { _ = try await startup.value }
            } else {
                try await runtimeFailure(mode == "error" ? "extension_start" : "browser_connection") { _ = try await startup.value }
            }
            try runtimeExpect(exits == 0, "Failed startup sent an active-session exit event.")
        }
        try runtimeExpect(Darwin.kill(pid, 0) != 0, "Browser test left a child running.")
        try runtimeExpect(try manager.contentsOfDirectory(atPath: data.path).isEmpty, "Browser test left a profile.")
    } catch {
        startup.cancel()
        if let running = try? await startup.value { await running.stop() }
        await browser?.stop()
        throw error
    }
}

@MainActor
func runRuntimeTests() async throws {
    let signalChild: EngineChildProcess = try {
        let originalHandler = signal(SIGTERM, SIG_IGN)
        defer { signal(SIGTERM, originalHandler) }
        var blocked = sigset_t(), originalMask = sigset_t()
        sigemptyset(&blocked)
        sigaddset(&blocked, SIGTERM)
        try runtimeExpect(pthread_sigmask(SIG_BLOCK, &blocked, &originalMask) == 0, "Could not set the fixture signal mask.")
        defer { pthread_sigmask(SIG_SETMASK, &originalMask, nil) }
        return try EngineChildProcess(executable: URL(fileURLWithPath: "/bin/sleep"), arguments: ["10"])
    }()
    signalChild.beginStop(grace: .milliseconds(100))
    let signalStatus = await signalChild.wait()
    try runtimeExpect(signalStatus.signalled && signalStatus.code == SIGTERM,
                      "Child inherited an ignored or blocked SIGTERM and required SIGKILL.")
    let output = try await BrowserRuntime.runSystemCheck("Test check", command: "/bin/sh", arguments: ["-c", "printf '152.0.7977.82\\n'"])
    try runtimeExpect(output == "152.0.7977.82", "System check changed stdout.")
    do {
        _ = try await BrowserRuntime.runSystemCheck("Chromium signature check", command: "/bin/sh", arguments: ["-c", "printf 'private stdout'; printf 'private stderr' >&2; exit 7"])
        throw EngineFailure("test", "Expected a failing system check.")
    } catch let error as EngineFailure {
        try runtimeExpect(error.code == "browser_verify" && error.message == "Chromium signature check failed (sh, exit 7).", "System check leaked output or lost its error context.")
    }
    try await runtimeFailure("browser_setup") {
        _ = try await BrowserRuntime.runSystemCheck("Chromium Gatekeeper check", command: "/bin/sleep", arguments: ["10"], timeout: .milliseconds(25))
    }
    let cancelled = Task { try await BrowserRuntime.runSystemCheck("Test check", command: "/bin/sleep", arguments: ["10"]) }
    try await Task.sleep(for: .milliseconds(25))
    cancelled.cancel()
    try await runtimeFailure("cancelled") { _ = try await cancelled.value }

    let checksum = "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad"
    try await runtimeDownloadTest { url, file in
        var progress: [BrowserRuntime.Status] = []
        try await BrowserRuntime.downloadArchive(from: url, to: file, checksum: checksum, progress: { progress.append($0) })
        try runtimeExpect(try Data(contentsOf: file) == Data("abc".utf8), "Download bytes changed.")
        let attributes = try FileManager.default.attributesOfItem(atPath: file.path)
        try runtimeExpect((attributes[.posixPermissions] as? NSNumber)?.intValue == 0o600, "Download file is not private.")
        try runtimeExpect(progress.first?.receivedBytes == 0 && progress.first?.fraction == 0,
                          "Download progress did not begin at zero.")
        try runtimeExpect(progress.last?.receivedBytes == 3 && progress.last?.fraction == 1,
                          "Download progress did not reach the full response length.")
        try runtimeExpect(progress.allSatisfy { $0.phase == .downloading && $0.totalBytes == 3 && $0.appPath == nil },
                          "Download progress used a wrong phase, total, or installed path.")
        let fractions = progress.compactMap(\.fraction)
        try runtimeExpect(fractions == fractions.sorted() && fractions.allSatisfy { (0...1).contains($0) },
                          "Download progress went backward or left its range.")
        let payload = progress.last!.value
        try runtimeExpect(payload["phase"] as? String == "downloading" && payload["receivedBytes"] as? Int64 == 3 && payload["totalBytes"] as? Int64 == 3,
                          "Download progress lost its structured fields.")
    }
    try await runtimeDownloadTest(body: "wrong archive") { url, file in
        try await runtimeFailure("browser_checksum") { try await BrowserRuntime.downloadArchive(from: url, to: file, checksum: checksum, progress: { _ in }) }
        try runtimeExpect(!FileManager.default.fileExists(atPath: file.path), "Checksum failure left an archive.")
    }
    try await runtimeDownloadTest(status: 503) { url, file in
        var progress: [BrowserRuntime.Status] = []
        try await runtimeFailure("browser_download") { try await BrowserRuntime.downloadArchive(from: url, to: file, checksum: checksum, progress: { progress.append($0) }) }
        try runtimeExpect(!FileManager.default.fileExists(atPath: file.path), "HTTP failure created an archive.")
        try runtimeExpect(progress.isEmpty, "HTTP failure reported successful download progress.")
    }
    try await runtimeDownloadTest { url, file in
        try Data("existing".utf8).write(to: file)
        try await runtimeFailure("browser_download") { try await BrowserRuntime.downloadArchive(from: url, to: file, checksum: checksum, progress: { _ in }) }
        try runtimeExpect(try String(contentsOf: file, encoding: .utf8) == "existing", "Download replaced an existing file.")
    }
    try await runtimeDownloadTest(body: String(repeating: "p", count: 8_192), hold: true) { url, file in
        let download = RuntimeDownloadTask()
        var progress: [BrowserRuntime.Status] = []
        download.task = Task {
            try await BrowserRuntime.downloadArchive(from: url, to: file, checksum: checksum, progress: {
                progress.append($0)
                if ($0.receivedBytes ?? 0) > 0 { download.task?.cancel() }
            })
        }
        try await runtimeFailure("cancelled") { try await download.task?.value }
        download.task = nil
        try runtimeExpect(!FileManager.default.fileExists(atPath: file.path), "Cancellation left a partial archive.")
        try runtimeExpect(progress.contains { ($0.receivedBytes ?? 0) > 0 } && progress.allSatisfy { ($0.fraction ?? 0) < 1 },
                          "Cancelled download did not report partial progress.")
    }
    var setupProgress: [BrowserRuntime.Status] = []
    let cancelledSetup = Task { try await BrowserRuntime.resolve(dataDirectory: FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString), progress: { setupProgress.append($0) }) }
    cancelledSetup.cancel()
    try await runtimeFailure("cancelled") { _ = try await cancelledSetup.value }
    try runtimeExpect(setupProgress.count == 1 && setupProgress[0].phase == .idle && setupProgress[0].message == nil,
                      "Cancelled setup reported a failure instead of idle.")
    let invalidDirectory = FileManager.default.temporaryDirectory.appendingPathComponent("passtrami-invalid-runtime-" + UUID().uuidString)
    try Data("fixture file".utf8).write(to: invalidDirectory)
    defer { try? FileManager.default.removeItem(at: invalidDirectory) }
    setupProgress.removeAll()
    try await runtimeFailure("browser_setup") {
        _ = try await BrowserRuntime.resolve(dataDirectory: invalidDirectory, progress: { setupProgress.append($0) })
    }
    try runtimeExpect(setupProgress.last?.phase == .failed && setupProgress.last?.message != nil && setupProgress.last?.appPath == nil,
                      "Setup failure did not report a separate runtime error.")
    for mode in ["ready", "wait-ready", "wait-cdp", "invalid", "oversized", "error", "exit", "spawn-error"] { try await runtimeBrowserTest(mode: mode) }
    var pipeFailed = false
    let pipe = try BrowserDebugPipe { pipeFailed = true }
    defer { pipe.close() }
    try await runtimeFailure("extension_start") {
        try await pipe.loadExtension(at: URL(fileURLWithPath: "/fixture"), timeout: .milliseconds(25))
    }
    try runtimeExpect(pipeFailed, "A pipe startup timeout did not stop the browser.")
    print("Runtime download, process, private CDP pipe, and cleanup checks passed.")
}
